lib/Dialect/FIRRTL/Transforms/InnerSymbolDCE.cpp

```diff
@@ -0,0 +1,122 @@
+//===- InnerSymbolDCE.cpp - Delete Unused Inner Symbols----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass removes inner symbols which have no uses.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SubElementInterfaces.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-inner-symbol-dce"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+using namespace hw;
+
+struct InnerSymbolDCEPass : public InnerSymbolDCEBase<InnerSymbolDCEPass> {
+  void runOnOperation() override;
+
+private:
+  void findInnerRefs(Attribute attr);
+  void insertInnerRef(InnerRefAttr innerRef);
+  void removeInnerSyms(FModuleOp op);
+
+  DenseSet<std::pair<StringAttr, StringAttr>> innerRefs;
+};
+
+/// Find all InnerRefAttrs inside a given Attribute.
+void InnerSymbolDCEPass::findInnerRefs(Attribute attr) {
+  // Check if this Attribute is an InnerRefAttr.
+  if (auto innerRef = dyn_cast<InnerRefAttr>(attr)) {
+    insertInnerRef(innerRef);
+    return;
+  }
+
+  // Check if any sub-Attributes are InnerRefAttrs.
+  if (auto subElementAttr = dyn_cast<SubElementAttrInterface>(attr))
+    subElementAttr.walkSubAttrs([&](Attribute subAttr) {
+      if (auto innerRef = dyn_cast<InnerRefAttr>(subAttr))
+        insertInnerRef(innerRef);
+    });
+}
+
+/// Add an InnerRefAttr to the set of all InnerRefAttrs.
+void InnerSymbolDCEPass::insertInnerRef(InnerRefAttr innerRef) {
+  StringAttr moduleName = innerRef.getModule();
+  StringAttr symName = innerRef.getName();
+
+  auto [iter, inserted] = innerRefs.insert({moduleName, symName});
+  if (!inserted)
+    return;
+
+  ++numSymbolsFound;
```

Comment:
This statistic seems to be counting unique InnerRef's encountered, which does seem useful but isn't what I'd expect for this stat based on its name/description.